apistar/server/app.py

```diff
@@ -16,37 +16,35 @@ def exception_handler(exc: Exception) -> Response:
 
 
 class App():
-    def __init__(self, document, components=None):
+    def __init__(self, document, template_dir=None, template_apps=None, components=None):
         components = list(components) if components else []
 
         self.document = document
-        self.components = self.default_components() + components
-        self.router = self.default_router()
-        self.templates = self.default_templates()
-        self.injector = self.default_injector()
+        self.router = self.init_router(document)
+        self.templates = self.init_templates(template_dir, template_apps)
+        self.injector = self.init_injector(components)
         self.exception_handler = exception_handler
 
-    def get_initial_components(self):
-        return {
+    def init_router(self, document):
+        return Router(document)
+
+    def init_templates(self, template_dir: str=None, template_apps: list=None):
+        return Templates(template_dir, template_apps, {
+            'reverse_url': self.reverse_url,
+            'static_url': self.static_url
+        })
+
+    def init_injector(self, components=None):
+        components = [] if (components is None) else list(components)
```

Comment:
why the parens around `(components is None)`?

also, small nit, on line 20 you use the style
```python
components = list(components) if components else []
```
why not normalize on one style and use in both places?

Reply:
Good call - I've gone with the simpler style.

> why the parens around (components is None)

I think `... if ... else ...` is easier to visually parse if any expression in the middle section is demarcated by being enclosed in parentheses. However we don't need that for single-variable expressions, so this is better anyways.